include/sound/sof/header.h

```diff
@@ -100,6 +101,16 @@
 #define SOF_IPC_STREAM_VORBIS_PARAMS		SOF_CMD_TYPE(0x010)
 #define SOF_IPC_STREAM_VORBIS_FREE		SOF_CMD_TYPE(0x011)
 
+/* probe */
+#define SOF_IPC_PROBE_INIT			SOF_CMD_TYPE(0x001)
+#define SOF_IPC_PROBE_DEINIT			SOF_CMD_TYPE(0x002)
+#define SOF_IPC_PROBE_DMA_SET			SOF_CMD_TYPE(0x003)
+#define SOF_IPC_PROBE_DMA_GET			SOF_CMD_TYPE(0x004)
+#define SOF_IPC_PROBE_DMA_DETACH		SOF_CMD_TYPE(0x005)
+#define SOF_IPC_PROBE_POINT_SET		SOF_CMD_TYPE(0x006)
+#define SOF_IPC_PROBE_POINT_GET		SOF_CMD_TYPE(0x007)
+#define SOF_IPC_PROBE_POINT_REMOVE		SOF_CMD_TYPE(0x008)
+
```

Comment:
This seems exaggerated to me. Does the kernel really have to control the DMA in the FW? Doesn't the FW know what to do once it gets a buffer ID to attach to? Why do we need GETs? Cannot the kernel keep all the information about the already configured probes? In principle, I think, we could also avoid INIT / DEINIT - the first probe initialises the last one removed frees, but ok, maybe explicit INIT / DEINIT makes it a bit easier for the FW. So, wouldn't the minimal set be just SET + REMOVE? You could even limit it to just a SET with an additional ON / OFF parameter, but I agree, that using a REMOVE makes it more explicit.

Reply:
I've explained in depth probes both here and in @akloniex PR regarding FW probes implementation. Given the sheer amount of questions, I'd like to propose a meeting with you and others and SOF FW members where we can explain the flow so it's all clear.


> This seems exaggerated to me. Does the kernel really have to control the DMA in the FW? Doesn't the FW know what to do once it gets a buffer ID to attach to?

One is completely disconnected from another. Probe point connection is dynamic and thus - except from extractor - FW needs explicit stream_tag input from user (kernel) in order to connect to the right endpoint (probe point).


 > Why do we need GETs? Cannot the kernel keep all the information about the already configured probes?

Without GETs you lose a lot of dynamism. I encourage you to look outside of SETs-only ties.
GETs allow me to fetch firmware AND hardware-implementation specifics before allocating resources (kernel driver probe).
GETs allow me to fetch module list descriptors so my topology graph knows its capabilities.
GETs allow me to retrieve list of runtime pipelines so userspace tools e.g.: FDK can expose them via GUI - client friendly interface.
<and the list goes on>

Keeping all the info inside kernel increases its resource use. In most cases, you don't need to know all the information all the time. You need them only for specific moment/ usecase.


> In principle, I think, we could also avoid INIT / DEINIT - the first probe initialises the last one removed frees, but ok, maybe explicit INIT / DEINIT makes it a bit easier for the FW. So, wouldn't the minimal set be just SET + REMOVE? You could even limit it to just a SET with an additional ON / OFF parameter, but I agree, that using a REMOVE makes it more explicit.

Having separate preparation step allows kernel to tell FW whether it needs extraction or not so HDAudio stream is not occupied unnecessarily. Without such, FW would always expect host to capture one of HDAudio streams for extraction, regardless if extraction is actually made use of or not.

Again, this is also connected to cAVS interface and probes are set to follow the cAVS path as closely as possible. Soon, Intel Audio Linux drivers stack will be switching to cAVS/ next gen model entirely with both, topology and IPC interface.

Traces, a probes close relative (debug wise), are a very similar subject, in regard to your INIT/ DEINIT question. Right now, SOF fires tracing right from the get-go, while in cAVS you enable them only when necessary - again, allowing host to save resources. To be explicit: that is done via ENABLE_LOGS(true/ false), similarly to probe's INIT/ DEINIT.